pull/github.go

```diff
@@ -829,6 +830,68 @@ func (ghc *GitHubContext) getCheckStatuses() (map[string]string, error) {
 	return statuses, nil
 }
 
+func (ghc *GitHubContext) LatestWorkflowRuns() (map[string]string, error) {
+	if ghc.workflowRuns != nil {
+		return ghc.workflowRuns, nil
+	}
+
+	opt := &github.ListWorkflowRunsOptions{
+		ExcludePullRequests: true,
+		HeadSHA:             ghc.HeadSHA(),
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+			Page:    0,
+		},
+	}
+
+	// The same workflow file can be triggered multiple times. For example:
+	//
+	// on:
+	//  pull_request:
+	//    types:
+	//      - opened
+	//      - synchronize
+	//      - edited
+	//
+	// If the workflow fails for the `opened` event, yet passes for `edited` and
+	// the user edits the PR, the conclusion of the workflow run for the
+	// `edited` event is the one we want to consider. If they then push another
+	// commit which fails then we should take the failure.
+	runsWithDate := make(map[string]*github.WorkflowRun)
+	for {
+		runs, resp, err := ghc.client.Actions.ListRepositoryWorkflowRuns(ghc.ctx, ghc.owner, ghc.repo, opt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get workflow runs for page %d", opt.Page)
+		}
+
+		for _, run := range runs.WorkflowRuns {
+			if run.GetStatus() != "completed" {
+				continue
+			}
+
+			// This is an older run than one we've already saw, so ignore it.
+			previousRun := runsWithDate[*run.Path]
+			if previousRun != nil && run.GetUpdatedAt().Before(previousRun.GetUpdatedAt().Time) {
+				continue
+			}
+
+			runsWithDate[*run.Path] = run
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+	workflowRuns := make(map[string]string, len(runsWithDate))
+
+	for path, run := range runsWithDate {
+		workflowRuns[path] = run.GetConclusion()
+	}
```

Comment:
That would be more aligned with what GitHub does. I constructed the situation in a test repo with a GitHub required status on `build b`. The `push` event fails and the `pull_request` one passes. The PR is blocked.

I'll do a change along the lines of ☝️ 

![image](https://github.com/user-attachments/assets/4f42cff4-ef18-49c7-9d43-ed17c6d440c0)

Reply:
Ran out of time to test this live today but I think the current version is more or less what I had in mind. I'll open up for review!

Reply:
Ok, I've now built an image with this change (and my other PRs) included and it seems to do what I expect

![image](https://github.com/user-attachments/assets/69719cb4-b0f4-427b-aa62-fbb224e4be31)

`policy-bot: main` is running the previous version of this change as per the first screenshot.

In this PR, the successful `pull_request` run for "Build directory B" came _after_ the failed `push` event run. In the previous version, since that came later, we considered it a success. Now, with the current version, we consider them separately since they are triggered by different event types, and the PR is blocked.

When the `push` workflow passes, we give a check as expected:

![image](https://github.com/user-attachments/assets/268e82c7-4d41-4c0c-bf70-69a3cf342ff9)